File: browser/src/js/moneysocket/stack/consumer.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const WebsocketLocation = require(
    '../beacon/location/websocket.js').WebsocketLocation;

const OutgoingWebsocketLayer = require(
    "../layer/outgoing_websocket.js").OutgoingWebsocketLayer;
const OutgoingRendezvousLayer = require(
    "../layer/outgoing_rendezvous.js").OutgoingRendezvousLayer;
const ConsumerLayer = require("../layer/consumer.js").ConsumerLayer;
const ConsumerTransactLayer = require(
    "../layer/consumer_transact.js").ConsumerTransactLayer;

class ConsumerStack {
    constructor() {
        this.onannounce = null;
        this.onrevoke = null;
        this.onproviderinfo = null;
        this.onstackevent = null;
        this.onping = null;
        this.oninvoice = null;
        this.onpreimage = null;

        this.websocket_layer = this.setupOutgoingWebsocketLayer(null);
        this.rendezvous_layer = this.setupOutgoingRendezvousLayer(
            this.websocket_layer);
        this.consumer_layer = this.setupConsumerLayer(this.rendezvous_layer);
        this.transact_layer = this.setupConsumerTransactLayer(
            this.consumer_layer);

        this.transact_layer.onannounce = (function(nexus) {
            this.announceNexus(nexus);
        }).bind(this);
        this.transact_layer.onrevoke = (function(nexus) {
            this.revokeNexus(nexus);
        }).bind(this);

        this.nexus = null;
    }


    //////////////////////////////////////////////////////////////////////////
    // setup
    //////////////////////////////////////////////////////////////////////////

    setupConsumerTransactLayer(below_layer) {
        var l = new ConsumerTransactLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("CONSUMER_TRANSACT", nexus, status);
        }).bind(this);
        l.oninvoice = (function(nexus, bolt11, request_reference_uuid) {
            this.onInvoice(nexus, bolt11, request_reference_uuid);
        }).bind(this);
        l.onpreimage = (function(nexus, preimage, request_reference_uuid) {
            this.onPreimage(nexus, preimage, request_reference_uuid);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupConsumerLayer(below_layer) {
        var l = new ConsumerLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("CONSUMER", nexus, status);
        }).bind(this);
        l.onproviderinfo = (function(nexus, msg) {
            this.onProviderInfo(nexus, msg);
        }).bind(this);
        l.onping = (function(nexus, msecs) {
            this.onPing(nexus, msecs);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupOutgoingRendezvousLayer(below_layer) {
        var l = new OutgoingRendezvousLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("OUTGOING_RENDEZVOUS", nexus, status);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupOutgoingWebsocketLayer() {
        var l = new OutgoingWebsocketLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("OUTGOING_WEBSOCKET", nexus, status);
        }).bind(this);
        return l;
    }

    //////////////////////////////////////////////////////////////////////////
    // transact layer callbacks:
    //////////////////////////////////////////////////////////////////////////

    onInvoice(transact_nexus, bolt11, request_reference_uuid) {
        if (this.oninvoice != null) {
            this.oninvoice(bolt11, request_reference_uuid);
        }
    }

    onPreimage(transact_nexus, preimage, request_reference_uuid) {
        if (this.onpreimage != null) {
            this.onpreimage(preimage, request_reference_uuid);
        }
    }

    //////////////////////////////////////////////////////////////////////////
    // consumer layer callbacks:
    //////////////////////////////////////////////////////////////////////////

    onProviderInfo(consumer_nexus, msg) {
        var provider_info = {'payer':         msg['payer'],
                             'payee':         msg['payee'],
                             'wad':           msg['wad'],
                             'account_uuid':  msg['account_uuid']};
        if (this.onproviderinfo != null) {
            this.onproviderinfo(provider_info);
        }
    }

    onPing(consumer_nexus, msecs) {
        if (this.onping != null) {
            this.onping(msecs);
        }
    }

    //////////////////////////////////////////////////////////////////////////
    // layer callbacks:
    //////////////////////////////////////////////////////////////////////////

    announceNexus(below_nexus) {
        this.nexus = below_nexus;
        this.shared_seed = below_nexus.getSharedSeed();

        if (this.onannounce != null) {
            this.onannounce(below_nexus);
        }
    }

    revokeNexus(below_nexus) {
        console.log("consumer stack got nexus revoked");
        this.nexus = null;
        this.shared_seed = null;

        if (this.onrevoke != null) {
            this.onrevoke(below_nexus);
        }
    }

    onLayerEvent(layer_name, nexus, status) {
        if (this.onstackevent != null) {
            this.onstackevent(layer_name, nexus, status);
        }
    }

    //////////////////////////////////////////////////////////////////////////
    // UI calls these
    //////////////////////////////////////////////////////////////////////////

    doConnect(beacon) {
        console.log("consumer stack connect called");

        var location = beacon.locations[0];
        var shared_seed = beacon.getSharedSeed();
        if (! (location instanceof WebsocketLocation)) {
            console.error("not websocket location?");
            return;
        }
        console.log("connect consumer: " + location.toWsUrl());
        this.websocket_layer.connect(location, shared_seed)
    }

    doDisconnect() {
        console.log("consumer stack disconnect called");
        this.websocket_layer.initiateCloseAll();
    }

    requestInvoice(msats, override_request_uuid, description) {
        this.nexus.requestInvoice(msats, override_request_uuid, description);
    }

    requestPay(bolt11, override_request_uuid) {
        this.nexus.requestPay(bolt11, override_request_uuid);
    }

}


exports.ConsumerStack = ConsumerStack;

File: python/moneysocket/stack/consumer.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

from moneysocket.beacon.location.websocket import WebsocketLocation
from moneysocket.layer.rendezvous.outgoing import OutgoingRendezvousLayer
from moneysocket.layer.websocket.outgoing import OutgoingWebsocketLayer
from moneysocket.layer.consumer import ConsumerLayer
from moneysocket.layer.transact.consumer import ConsumerTransactLayer


class ConsumerStack(object):
    def __init__(self):
        self.onannounce = None
        self.onrevoke = None
        self.onstackevent = None
        self.onproviderinfo = None
        self.onping = None
        self.oninvoice = None
        self.onpreimage = None

        assert self.rendezvous_layer # setup in subclass
        self.consumer_layer = self.setup_consumer_layer(self.rendezvous_layer)
        self.transact_layer = self.setup_transact_layer(self.consumer_layer)

    ###########################################################################

    def setup_consumer_layer(self, below_layer):
        l = ConsumerLayer()
        l.register_above_layer(below_layer)
        l.register_layer_event(self.send_stack_event, "CONSUMER")
        l.onping = self.on_ping
        return l

    def setup_transact_layer(self, below_layer):
        l = ConsumerTransactLayer()
        l.register_above_layer(below_layer)
        l.register_layer_event(self.send_stack_event, "CONSUMER_TRANSACT")
        l.oninvoice = self.on_invoice
        l.onpreimage = self.on_preimage
        l.onproviderinfo = self.on_provider_info
        return l

    ############# transact layer callbacks

    def on_invoice(self, transact_nexus, bolt11, request_reference_uuid):
        if self.oninvoice:
            self.oninvoice(transact_nexus, bolt11, request_reference_uuid)

    def on_preimage(self, transact_nexus, preimage, request_reference_uuid):
        if self.onpreimage:
            self.onpreimage(transact_nexus, preimage, request_reference_uuid)

    ############# consumer layer callbacks

    def on_provider_info(self, transact_nexus, msg):
        provider_info = {'payer':         msg['payer'],
                         'payee':         msg['payee'],
                         'wad':           msg['wad'],
                         'account_uuid':  msg['account_uuid']}
        if self.onproviderinfo:
            self.onproviderinfo(transact_nexus, provider_info)

    def on_ping(self, transact_nexus, msecs):
        if self.onping:
            self.onping(msecs)

    ######## layer callback

    def on_announce(self, below_nexus):
        if self.onannounce:
            self.onannounce(below_nexus);

    def on_revoke(self, below_nexus):
        if self.onrevoke:
            self.onrevoke(below_nexus)


    def send_stack_event(self, layer_name, nexus, status):
        if self.onstackevent:
            self.onstackevent(layer_name, nexus, status)

    ######## UI calls these

    def do_connect(self, beacon):
        # implement in subclass
        pass

    def do_disconnect(self):
        # implement in subclass
        pass

    def request_invoice(self, nexus_uuid, msats, description):
        return self.transact_layer.request_invoice(nexus_uuid, msats,
                                                   description)

    def request_pay(self, nexus_uuid, bolt11):
         return self.transact_layer.request_pay(nexus_uuid, bolt11)


class OutgoingConsumerStack(ConsumerStack):
    def __init__(self):
        self.websocket_layer = self.setup_websocket_layer()
        self.rendezvous_layer = self.setup_rendezvous_layer(
            self.websocket_layer)
        super().__init__()

    ###########################################################################

    def setup_rendezvous_layer(self, below_layer):
        l = OutgoingRendezvousLayer()
        l.register_above_layer(below_layer)
        l.register_layer_event(self.send_stack_event, "OUTGOING_RENDEZVOUS")
        return l

    def setup_websocket_layer(self):
        l = OutgoingWebsocketLayer()
        l.register_layer_event(self.send_stack_event, "OUTGOING_WEBSOCKET")
        return l

    ###########################################################################

    def do_connect(self, beacon):
        location = beacon.locations[0]
        shared_seed = beacon.get_shared_seed()
        if type(location) != WebsocketLocation:
            return None;
        return self.websocket_layer.connect(location, shared_seed)

    def do_disconnect(self):
        self.websocket_layer.initiate_close_all()


# TODO incoming websocket stack
